server.go

```diff
@@ -466,6 +476,9 @@ Loop:
 // shutdown closes current clients and also stops accepting new clients in a
 // non-blocking manner
 func (s *Server) shutdown() {
+	// Stop accepting new consumers while a server shutdown operation is in-progress.
+	s.teardown = true
```

Comment:
There is a possible data race here. Since `s.teardown` may be accessed concurrently by `shutdown()` and `Handle()`, it should be guarded with a lock. In this particular case, see [`RWMutex`](https://golang.org/pkg/sync/#RWMutex).

Reply:
Since `Handle` only reads the value to check if it should consume or not we should probably decide if we want any [performance regression](https://github.com/golang/go/issues/17973) that might get introduced by this mutex or we can live with having a consumer trying to consume after we initiate a `teardown`.

Reply:
Interesting case. We could benchmark the performance regression of maintaining a mutex in this variable and decide accordingly IMHO. I suggest to better handle this in a separate PR and create an issue to not forget it. What do you think @agis, @0xdiba ?

Reply:
Synchronization is needed here, otherwise it means consumers might be created *after* `shutdown()` is called, which can lead to all sorts of undefined behavior (since we're dealing with librdkafka under the hood). Best-case scenario, consumers may be outstanding and not finalized (commit offsets etc.) when rafka shuts down.

Regarding the referenced upstream issue, it won't affect us since the cost of calling RWLock/RWUnlock is amortized by the work `Consume()` is doing. It would be an issue if we had a hot-path calling RWLock very fast, but in rafka this is not an issue.

That said, if RWMutex's performance was a concern, we could use `sync.Atomic` or channels or something else.

Reply:
Agreed, I'll proceed as you suggest.